_posts/2018-01-05-python101.md

```diff
@@ -0,0 +1,847 @@
+
+### A Short Detour
+You might be wondering what the ``self`` variable means. It's known as an **implicit parameter**, which means that when calling the function, you specify the object name _before_ the function name. Here's a concrete example to clear things up.
+
+```python
+class Table:
+  def __init__(self):
+    self.height = 10
+
+  def change_height(self, amount):
+    self.height += amount
+
+t = Table() # create an instance of the Table class, aka create a Table object
+t.change_height(10) # call change_height
+Table.change_height(t, 10) # does the same thing as the previous line
+```
+
+Say that we have an object named ``t``, of class type ``Table``. Suppose that ``Table`` has a function ``change_height``. We can thus call ``change_height`` two different ways - by implicitly passing the parameter, or explicitly passing the parameter. If we call ``change_height`` like ``t.change_height(10)``, then ``t``, our object, is named _before_ the function name, ``change_height``. Thus, ``change_height`` was called with an implicit parameter. Alternatively, we could have written ``Table.change_height(t, 10)``, where we pass in the object **explicitly**.
+
+### Back to the Gradebook Application
+
+Now that we have the ``Student`` class written, it's time to start writing the ``Course`` class, which will represent the courses that students are taking. Each ``Course`` should have a course name, a list of enrolled students, and a list of assignments, so we'll create those as **instance variables**. Additionally, we need ways to add and remove assignments, as well as set and get grades for the students, so we'll create those as functions.
+
+```python
+class Course:
+  def __init__(self, name):
+    self.course_name = name
+    self.enrolled_students = []
+    # assignments is a dictionary with key=assignment name, value=dictionary of students
+    #   with key=Student object, value=assignment grade
+    self.assignments = {}
+
+  def add_assignment(self, assignment):
+    self.assignments[assignment] = {}
+
+  def remove_assignment(self, assignment):
+    del self.assignments[assignment]
+
+  def set_grade(self, assignment, student, grade):
+    self.assignments[assignment][student] = grade # assign a grade to a Student
+
+  def get_overall_grade_for(self, student):
+    sum_scores = 0  # need an accumulator (see Glossary) to compute average
+    num_graded = 0  # need to keep track of total number of grades for average
+    for assignment, grades in self.assignments.items(): # iterate over all items in self.assignments
+      if student in grades.keys():  # look to see if the student has a grade
+        sum_scores += grades[student]  # add that grade to the accumulator
+        num_graded += 1 # increment number of graded assignments (the denominator)
```

Comment:
if you're going do all this dictionary stuff in the example for classes, you'll need to fill in that dictionaries section above